### DISCUSSION TITLE
Testes de performance

### DISCUSSION THREAD

**Discussion by dionarodrigues** [frontendbr/forum#43] [Perguntas e Respostas]
Galera, performance sempre foi um assunto muito importante na web, principalmente agora com os mobiles e outros diversos dispositivos. Por isso, criei essa issue para saber como vocês têm feito para realizar os testes: há alguma ferramenta, workflow ou checklist que vocês gostam e indicam?

---

Eu, por exemplo, sempre me baseio nas dicas do [Browser Diet](https://browserdiet.com/pt/) como princípios básicos. Também utilizo a [ferramenta Page Speed do Google](https://developers.google.com/speed/pagespeed/insights/) para identificar pontos críticos que precisam ser melhorados.

Recentemente descobri a ferramenta [GTmetrix](https://gtmetrix.com/features.html), que faz uma análise do site utilizando diversas ferramentas, como o PageSpeed, Yahoo! Yslow, dentre muitas outras possibilidades; e a ferramenta [Selenium IDE](https://addons.mozilla.org/pt-br/firefox/addon/selenium-ide/), que iremos começar a utilizar aqui na empresa que trabalho para gravar, editar e depurar os testes via browser.

**Comment by karlprieb** [comment]
Acho que devemos ir além destas análises e também partir para desempenho CSS e JS. Atualmente tanto no Web Tools do Chrome quanto do Firefox existem ferramentas decentes para fazer esse tipo de análise :)

Cuidado com o GTmetrix, eu utilizava ele até perceber que os resultados são divergentes do Page Speed do Google.

**Comment by tcelestino** [comment]
Semana passada assistir esse [video](https://www.youtube.com/watch?v=O3W1yuq-ZlE&feature=youtu.be) bem antigo do Paul Irish mostrando como trabalhar com o DevTools, fiquei impressionando como a ferramenta é útil.

**Comment by felipefialho** [comment]
Sempre me baseei muito no GT Metrix para ter uma ideia de como anda a performance do projeto. @karlprieb Os resultados são divergentes? Dessa não sabia.

O Selenium apesar de ser um projeto já antigo e com poucas atualizações, continua justo, acredito que não tem nada melhor que ele para essa função.

**Comment by fdaciuk** [comment]
Só pra tretar com o Selenium: [DalekJS](http://dalekjs.com/) é mais legal :stuck_out_tongue:

**Comment by karlprieb** [comment]
@LFeh Só do PageSpeed. O YSlow não tem divergências.

**Comment by dionarodrigues** [comment]
Que bad notice @karlprieb sobre o GTmetrix. Pode dar alguns exemplos das inconsistências que você identificou? 

Boa @tcelestino . Vou assistir o vídeo pra conseguir aproveitar mais o DevTools man. ;)

Mamílico heim @fdaciuk . O DalekJS parece ser uma solução mais moderna por um lado, mas faltou a famosa IDE do Selenium. =/ HAuahuhahua

**Comment by fdaciuk** [comment]
aeuhae! Mas a ideia do Dalek é realmente não usar Selenium. Praticamente todas as ferramentas de testes E2E usam Selenium por trás, menos o Dalek, que por sinal é muito mais performático ;)

**Comment by felipefialho** [comment]
Não conhecia @fdaciuk, valeu por compartilhar ;)

**Comment by rubenmarcus** [comment]
e o Pingdom?

**Comment by dionarodrigues** [comment]
@fdaciuk kara, tirei essa manhã pra fazer testes com o DalekJS. Sensacional!

É muito fácil de se usar, configurar e realizar os testes em diferentes browsers. Cria-se as tarefas com JS, pode ser integrado com o Grunt, gera relatórios flexíveis.... Ou seja, perfect!

Contudo, conversando com o pessoal de TI aqui na empresa, chegamos à conclusão que, para testes rápidos e que podem ser feitos por qualquer pessoa (já que não temos testers aqui), é interessante utilizar o Selenium (já sabendo que os testes serão feitos apenas no firefox). Mas, para testes de Front-end, pode ser interessante adicionar alguns testes com o DalekJS no workflow. Como por exemplo, testar diversas ações na interface em tamanhos diversos e browsers diferentes para saber se nada está quebrando....

Então, o que você acha disso? Em quais situações voltadas para o front-end você costuma utilizar ele?

**Comment by rickbenetti** [comment]
@rubenmarcus uns anos atrás um cara que cuidava do server de um cliente meu disse que não confiava no pingdom e sites do tipo pois eles apresentam dados sempre ruim, pois o objetivo deles é que você contrate a solução deles, tipo os sites na pingdom não tem referência de comparativo com o o gtmetrix faz por exemplo.

Eu ainda uso, mas não como muita base, serve para ter uma leve noção hoje em dia.

**Comment by karlprieb** [comment]
@diogorodrigues 
GTMetrix: http://s12.postimg.org/eojdrvse5/gtmetrix.png
PageSpeed: http://s12.postimg.org/hhclbwsql/pagespeed.png

Resultados diferentes

**Comment by fdaciuk** [comment]
> Contudo, conversando com o pessoal de TI aqui na empresa, chegamos à conclusão que, para testes rápidos e que podem ser feitos por qualquer pessoa (já que não temos testers aqui), é interessante utilizar o Selenium

@diogorodrigues, então, acho que depende do conhecimento da equipe. Dalek é JS. As duas ferramentas foram feitas para testes no browser mesmo (end to end). Mas se o time está mais acostumado com Java, use o Selenium. Se já conhecem JS, eu iria de Dalek.

A grande vantagem que eu vejo no Dalek é:
1. Eu consigo usar um browser headless (sem janela - Phantom);
2. A API dele é em Node, mas com o método [`.execute`](http://dalekjs.com/docs/actions.html#meth-execute), você consegue executar código JS no contexto do browser. A API é em Node, mas todo o código dentro de `.execute` vai rodar no browser, permitindo usar `window`, `document`, ou qualquer outra coisa que eu precise, que esteja no escopo do browser. Para fazer isso com Selenium, você precisa escrever uma String ;)
3. Pela vantagem de poder usar headless browser, e a api ser em Node, eu posso facilmente usar ele em um CI/CD, para garantir que não estou entregando bugs em produção.

Enfim, tem mais vantagens, mas acho que esses são os pontos mais importantes :)

**Comment by vitorbritto** [comment]
Fiz um _update_ pesado nos últimos meses para os meus testes. Deixo a lista completa. Alguma destas ferramentas pode ser útil para outro colega.
- Testes de **Performance**: [JSPerf](http://jsperf.com/), Dev tools do Firefox, [OpenBeat](http://opbeat.com), [New Relic](http://newrelic.com)
- Testes de **Carga/Stress**: [Gremlins](https://github.com/marmelab/gremlins.js)
- Testes de **Integração**: [Pioneer](http://pioneerjs.com/)
- Testes de **E2E**: [DalekJS](http://dalekjs.com/)
- Testes de **Regressão**: [BackToStopJS](https://garris.github.io/BackstopJS/)

**Comment by felipefialho** [comment]
Perfeito @vitorbritto!

**Comment by mksanti** [comment]
@vitorbritto 👍🏻

**Comment by vitor-mendrone-deel** [comment]
@vitorbritto cara eu tava precisando muito de algo tipo o Gremlins! Obrigado por compartilhar! :grin:

**Comment by lagden** [comment]
@vitorbritto Pioneerr!!! Eu não tinha colocado no favoritos e fiquei que nem louco procurando no Google plus!!! :metal: :godmode:

**Comment by vitorbritto** [comment]
@Mendrone excelente ferramenta! E divertida também, hehehe! :ghost:

**Comment by vitorbritto** [comment]
@lagden gostei do Pioneer logo de cara por causa do Cucumber e da sintaxe Gherkin. 

Cada vez mais utilizo testes de integração no lugar de testes unitários. Claro que não é possível utilizar em todos os cenários, mas dou prioridade a testes de integração sempre que possível. =]

**Comment by victormiguez** [comment]
@vitorbritto muito obrigado, cara! estava procurando exatamente por isso, links devidamente salvos :)
Só um detalhe, o link do OpenBeat está errado, apontando para "https://github.com/frontendbr/forum/issues/opbeat.com".

**Comment by fdaciuk** [comment]
Ajustei os links @victormiguez, valeu! :+1: 

/cc @vitorbritto :D

**Comment by vitorbritto** [comment]
@victormiguez opa! Disponha meu caro. =]

@fdaciuk obrigado! :ghost:

**Comment by lnfnunes** [comment]
@vitorbritto valeu pelos links!
Por favor, poderia me explicar melhor qual a diferença (prática), entre:
- Testes de Integração: Pioneer
- Testes de E2E: DalekJS

Me pareceu que com qualquer 1 dos dois eu consigo atingir o mesmo objetivo, então fiquei com essa dúvida do pq vc separou os dois em ferramentas diferentes.
Seria a questão das sintaxes ou envolve outros detalhes? Como vc utiliza no seu workflow?
Tks

**Comment by dionarodrigues** [comment]
Ótimas dicas galera. Valeu @fdaciuk e @vitorbritto por compartilhar essas ferramentas. =D

**Comment by vitorbritto** [comment]
@lnfnunes abri a _issue_ #65 para discutirmos sobre Testes de Integração e Aceitação. Te respondi por lá. =]

**Comment by lagden** [comment]
@vitorbritto 
Uma dúvida!!

Lá em cima você escreveu:

> Testes de Integração: Pioneer
> Testes de E2E: DalekJS

O `Pioneer` e o `DalekJS` não tem o mesmo propósito, mas com features diferentes?

**Comment by filipelinhares** [comment]
Tem uma ferramenta que gosto bastante que é o [Nightmare](https://github.com/segmentio/nightmare). O mais legal é a [chrome extension](https://github.com/segmentio/daydream) :smile:
